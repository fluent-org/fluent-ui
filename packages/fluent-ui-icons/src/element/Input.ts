export default {
  type: 'svg',
  key: null,
  ref: null,
  props: {
    version: '1.1',
    xmlns: 'http://www.w3.org/2000/svg',
    width: '32',
    height: '32',
    viewBox: '0 0 32 32',
    children: {
      type: 'path',
      key: null,
      ref: null,
      props: {
        d:
          'M6 12h-2v-2h2v2zM4 14h4v2h-4v-2zM22 14h4v2h-4v-2zM22 18h4v2h-4v-2zM4 18h4v2h-4v-2zM10 18h10v2h-10v-2zM10 12h-2v-2h2v2zM10 14h2v2h-2v-2zM14 14h2v2h-2v-2zM18 14h2v2h-2v-2zM14 12h-2v-2h2v2zM18 12h-2v-2h2v2zM22 12h-2v-2h2v2zM28.625 0q0.703 0 1.313 0.266t1.070 0.727 0.727 1.070 0.266 1.313v5.625q0 1.188-0.523 2.234t-1.477 1.766v9q0 0.422-0.156 0.781t-0.43 0.633-0.633 0.43-0.781 0.156h-26q-0.422 0-0.781-0.156t-0.633-0.43-0.43-0.633-0.156-0.781v-14q0-0.422 0.156-0.781t0.43-0.633 0.633-0.43 0.781-0.156h20v-2.625q0-0.703 0.266-1.313t0.727-1.070 1.070-0.727 1.313-0.266h3.25zM28 13.906q-0.5 0.094-1 0.094-1.031 0-1.945-0.391t-1.594-1.070-1.070-1.594-0.391-1.945v-1h-20v14h26v-8.094zM30 3.375q0-0.578-0.398-0.977t-0.977-0.398h-0.625v4h-2v-4h-0.625q-0.578 0-0.977 0.398t-0.398 0.977v5.625q0 0.625 0.234 1.172t0.641 0.953 0.953 0.641 1.172 0.234 1.172-0.234 0.953-0.641 0.641-0.953 0.234-1.172v-5.625z'
      },
      _owner: null,
      _store: {}
    }
  },
  _owner: null,
  _store: {}
}
