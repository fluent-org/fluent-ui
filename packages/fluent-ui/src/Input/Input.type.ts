import * as React from 'react'
import * as PropTypes from 'prop-types'
import { StandardProps } from '..'
import TextArea from './components/TextArea'

export type InputClassProps =
  | 'root'
  | 'error'
  | 'wrapper'
  | 'clearedIcon'
  | 'prefix'
  | 'suffix'
  | 'ghost'

export interface InputProps
  extends StandardProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement,
    'onChange' | 'prefix'
  > {
  value?: string
  onChange?: (value: string) => void
  placeholder?: string
  disabled?: boolean
  cleared?: boolean
  password?: boolean
  error?: boolean
  prefix?: React.ReactNode
  suffix?: React.ReactNode
  ghost?: boolean
}

export const InputPropTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  disabled: PropTypes.bool,
  cleared: PropTypes.bool,
  password: PropTypes.bool,
  error: PropTypes.bool,
  prefix: PropTypes.node,
  suffix: PropTypes.node,
  ghost: PropTypes.bool
}

export interface InputType extends React.ForwardRefExoticComponent<InputProps> {
  TextArea: typeof TextArea
}
