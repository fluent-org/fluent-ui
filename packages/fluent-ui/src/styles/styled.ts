import styled from 'styled-components'
import { th, variant, system, breakpoints } from '@xstyled/system'

export { styled, system, th, variant, breakpoints }

export * from 'styled-components'
export * from '@xstyled/system'
// export * from '@xstyled/prop-types'

export * from './stylize'
