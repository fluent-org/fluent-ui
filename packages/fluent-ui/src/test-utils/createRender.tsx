import * as React from 'react'
import * as PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

interface WrapperProps {
  children: React.ReactNode
}

const Wrapper: React.FC<WrapperProps> = ({ children }): React.ReactElement => (
  <ThemeProvider theme={{}}>{children}</ThemeProvider>
)

Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui: React.ReactElement, options: any) =>
  render(ui, { wrapper: Wrapper, ...options })

export { customRender as render }
