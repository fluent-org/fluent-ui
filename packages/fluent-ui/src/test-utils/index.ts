export * from '@testing-library/react'
// @ts-ignore
export * from './createRender'
