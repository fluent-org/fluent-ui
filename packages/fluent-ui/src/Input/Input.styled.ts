import { Style, Styles } from 'jss'
import { Theme } from '../styles'
import { InputClassProps } from './Input.type'

const wrapper: Style = {
  display: 'inline-block',
  position: 'relative',
  minWidth: 200,
  font: 'inherit'
}

const root = (theme: Theme) => ({
  outline: 'none',
  font: 'inherit',
  width: '100%',
  borderRadius: 2,
  border: '2px solid',
  borderColor: theme.colors!.standard!.default,
  transition: theme.transitions!.input,
  // ${th.size('medium.input')},
  '&:hover': {
    borderColor: theme.colors!.standard!.dark1
  },
  '&:active, &:focus': {
    borderColor: theme.colors!.primary!.default
  },
  '&:disabled': {
    color: theme.colors!.standard!.dark2,
    backgroundColor: theme.colors!.standard!.light1,
    cursor: 'not-allowed',
    pointerEvents: 'none'
  },
  ...theme.sizes!.medium!.input
})
const error = (theme: Theme): Style => ({
  borderColor: theme.colors!.error!.default,
  '&:hover': {
    borderColor: theme.colors!.error!.default
  },
  '&:active, &:focus': {
    borderColor: theme.colors!.error!.default
  }
})

const clearedIcon = (theme: Theme): Style => ({
  height: '100%',
  position: 'absolute',
  right: 0,
  top: 0,
  fontSize: 12,
  cursor: 'pointer',
  display: 'inline-flex',
  justifyContent: 'center',
  alignItems: 'center',
  transition: theme.transitions!.input,
  '&:hover': {
    backgroundColor: theme.colors!.primary!.default,
    color: theme.colors!.white!.default
  }
})

export const styles = (theme: Theme): Styles<InputClassProps> => ({
  root: root(theme),
  error: error(theme),
  wrapper,
  clearedIcon: clearedIcon(theme)
})
