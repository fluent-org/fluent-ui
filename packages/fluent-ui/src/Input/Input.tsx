import * as React from 'react'
import { StyledWrapper, StyledInput, StyledIcon } from './Input.styled'
import { ChromeClose as ChromeCloseIcon } from '@fluent-ui/icons'

export interface InputProps
  extends Omit<
    React.InputHTMLAttributes<HTMLInputElement>,
    'size' | 'value' | 'onChange' | 'type'
  > {
  value?: string
  onChange?: (value: string) => void
  placeholder?: string
  disabled?: boolean
  cleared?: boolean
  password?: boolean
}

const Input = React.forwardRef<HTMLInputElement, InputProps>(
  (
    { value, onChange, placeholder, disabled, cleared, password, ...rest }: InputProps,
    ref
  ): React.ReactElement => {
    function handleChange(e: React.ChangeEvent<HTMLInputElement> | null): void {
      onChange && onChange(e ? e.target.value : '')
    }
    function handleClear(e: React.MouseEvent<HTMLElement>): void {
      e.preventDefault()
      e.stopPropagation()
      handleChange(null)
    }
    return (
      <StyledWrapper>
        <StyledInput
          ref={ref}
          value={value}
          onChange={handleChange}
          placeholder={placeholder}
          disabled={disabled}
          cleared={cleared}
          type={password ? 'password' : 'text'}
          {...rest}
        />
        {value && cleared && (
          <StyledIcon onClick={handleClear}>
            <ChromeCloseIcon />
          </StyledIcon>
        )}
      </StyledWrapper>
    )
  }
)

Input.displayName = 'FInput'

export default Input
