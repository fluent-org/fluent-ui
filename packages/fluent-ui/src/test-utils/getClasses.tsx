import * as React from 'react'
import { createUseStyles } from '@fluent-ui/styles'
import { Styles } from 'jss'
import { render } from './createRender'

export function getClasses<T extends string>(styles: Styles, name: string): Record<T, string> {
  const useStyles = createUseStyles<T>(styles, { name })
  let classes: Record<T, string>
  function Util(props: any): React.ReactElement {
    classes = useStyles(props)
    return <div />
  }
  render(<Util />)
  // @ts-ignore
  return classes
}
