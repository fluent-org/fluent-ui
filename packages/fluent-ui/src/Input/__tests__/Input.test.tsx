import * as React from 'react'
import { fireEvent } from '@testing-library/react'
import { getClasses, render } from '../../test-utils'
import '@testing-library/jest-dom/extend-expect'

import Input, { name } from '../Input'
import { styles } from '../Input.styled'
import { InputClassProps } from '../Input.type'

const classes = getClasses<InputClassProps>(styles, name)

describe('Input', (): void => {
  const text = 'Hint Text'

  test('should be support ref', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    const { getByTestId } = render(<Input ref={ref} data-testid="input" />)
    const { current: element } = ref
    expect(element).toEqual(getByTestId('input'))
  })

  test('should be controlled', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    let value = ''
    function handleChange(v: string): void {
      value = v
    }
    const { container } = render(<Input ref={ref} value={value} onChange={handleChange} />)
    fireEvent.change(ref.current!, { target: { value: text } })
    expect(value).toEqual(text)
    expect(container.firstChild).toHaveClass(classes.wrapper)
    expect(ref.current!).toHaveClass(classes.root)
  })

  test('should be support placeholder', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    render(<Input ref={ref} placeholder={text} />)
    expect((ref.current as HTMLInputElement).placeholder).toEqual(text)
  })

  test('should be support disabled', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    render(<Input ref={ref} disabled={true} />)
    expect((ref.current as HTMLInputElement).disabled).toEqual(true)
  })

  test('should be support cleared', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    let value = ''
    function handleChange(v: string): void {
      value = v
    }
    const { container, sheets } = render(
      <Input ref={ref} value={value} onChange={handleChange} cleared={true} />
    )

    expect(container.firstChild).toMatchSnapshot()
    expect(sheets.toString()).toMatchSnapshot()

    fireEvent.change(ref.current!, { target: { value: text } })
    expect(value).toEqual(text)

    fireEvent.click(container.querySelector('.FInput-clearedIcon')!)
    expect(value).toEqual('')
  })

  test('should be support password', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    render(<Input ref={ref} password={true} />)
    expect((ref.current as HTMLInputElement).type).toEqual('password')
  })

  test('should be support error', (): void => {
    const ref = React.createRef<HTMLInputElement>()
    const { sheets } = render(<Input ref={ref} error={true} />)
    expect(ref.current as HTMLInputElement).toHaveClass(classes.error)
    expect(sheets.toString()).toMatchSnapshot()
  })
})
