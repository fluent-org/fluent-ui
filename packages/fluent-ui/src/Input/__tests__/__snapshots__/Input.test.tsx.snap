// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input prop error 1`] = `
".FInput-root {
  font: inherit;
  width: 100%;
  border: 2px solid;
  outline: none;
  padding: 6px 12px;
  transition: background-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,border-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms;
  border-color: #dedede;
  border-radius: 2px;
}
.FInput-root:hover {
  border-color: #b5b5b5;
}
.FInput-root:active, .FInput-root:focus {
  border-color: #0078d4;
}
.FInput-root:disabled {
  color: #8c8c8c;
  cursor: not-allowed;
  pointer-events: none;
  background-color: #e9e9e9;
}
.FInput-error {
  border-color: #f44336;
}
.FInput-error:hover {
  border-color: #f44336;
}
.FInput-error:active, .FInput-error:focus {
  border-color: #f44336;
}
.FInput-wrapper {
  font: inherit;
  display: inline-block;
  position: relative;
  min-width: 200px;
}
.FInput-clearedIcon {
  top: 0;
  right: 0;
  cursor: pointer;
  height: 100%;
  display: inline-flex;
  position: absolute;
  font-size: 12px;
  transition: background-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,border-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms;
  align-items: center;
  justify-content: center;
}
.FInput-clearedIcon:hover {
  color: #ffffff;
  background-color: #0078d4;
}"
`;

exports[`Input should be support cleared 1`] = `
<div
  class="FInput-wrapper"
>
  <input
    class="FInput-root"
    style="padding-right: 0px;"
    type="text"
    value=""
  />
  <span
    class="FInput-clearedIcon FTransition-root FTransition-root-0 FTransition-fade"
    style="width: 0px;"
  >
    <svg
      class="sc-kQsIoO elwqFk"
      viewBox="0 0 2048 2048"
    >
      <path
        d="M1169 1024l879 879-145 145-879-879-879 879L0 1903l879-879L0 145 145 0l879 879L1903 0l145 145z"
      />
    </svg>
  </span>
</div>
`;

exports[`Input should be support cleared 2`] = `
".FTransition-root {}
.FTransition-fade {}
.FTransition-fade.fade-enter {
  opacity: 0;
}
.FTransition-fade.fade-exit {
  opacity: 1;
}
.FTransition-fade.fade-exit-active {
  opacity: 0;
}
.FTransition-fade.fade-enter-active {
  opacity: 1;
}
.FTransition-zoom {}
.FTransition-zoom.zoom-enter {}
.FTransition-zoom.zoom-exit {}
.FTransition-zoom.zoom-exit-active {
  transform: scale3d(0, 0, 0);
}
.FTransition-zoom.zoom-exit-done {
  transform: scale3d(0, 0, 0);
}
.FTransition-zoom.zoom-enter-active {
  transform: scale3d(0, 0, 0);
}
.FTransition-zoom.zoom-enter-done {
  transform: none;
}
.FTransition-slide {}
.FTransition-slide.slide-enter {
  transform: translateY(100vh);
}
.FTransition-slide.slide-exit {
  transform: none;
}
.FTransition-slide.slide-exit-active {
  transform: translateY(100vh);
}
.FTransition-slide.slide-exit-done {
  transform: translateY(100vh);
}
.FTransition-slide.slide-enter-done {
  transform: none;
}
.FTransition-collapse {}
.FTransition-grow {}
.FTransition-grow.grow-enter {
  opacity: 0;
  transform: scale(0.75, 0.5625);
}
.FTransition-grow.grow-exit {}
.FTransition-grow.grow-exit-active {
  opacity: 0;
  transform: scale(0.75, 0.5625);
}
.FTransition-grow.grow-exit-done {
  opacity: 0;
  transform: scale(0.75, 0.5625);
}
.FTransition-grow.grow-enter-active {
  opacity: 1;
  transform: scale(1, 1);
}
.FTransition-grow.grow-enter-done {
  opacity: 1;
  transform: none;
}
.FTransition-root-0 {
  opacity: 0;
  transition: opacity 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,visibility 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms;
  visibility: hidden;
}
.FTransition-collapse-1 {}
.FTransition-collapse-1.collapse-enter {
  height: 0;
}
.FTransition-collapse-1.collapse-exit {
  height: 0;
}
.FTransition-collapse-1.collapse-exit-active {
  height: 0;
}
.FTransition-collapse-1.collapse-enter-active {
  height: 0;
}
.FTransition-collapse-1.collapse-enter-done {
  height: auto;
}
.FInput-root {
  font: inherit;
  width: 100%;
  border: 2px solid;
  outline: none;
  padding: 6px 12px;
  transition: background-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,border-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms;
  border-color: #dedede;
  border-radius: 2px;
}
.FInput-root:hover {
  border-color: #b5b5b5;
}
.FInput-root:active, .FInput-root:focus {
  border-color: #0078d4;
}
.FInput-root:disabled {
  color: #8c8c8c;
  cursor: not-allowed;
  pointer-events: none;
  background-color: #e9e9e9;
}
.FInput-error {
  border-color: #f44336;
}
.FInput-error:hover {
  border-color: #f44336;
}
.FInput-error:active, .FInput-error:focus {
  border-color: #f44336;
}
.FInput-wrapper {
  font: inherit;
  display: inline-block;
  position: relative;
  min-width: 200px;
}
.FInput-clearedIcon {
  top: 0;
  right: 0;
  cursor: pointer;
  height: 100%;
  display: inline-flex;
  position: absolute;
  font-size: 12px;
  transition: background-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,border-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms;
  align-items: center;
  justify-content: center;
}
.FInput-clearedIcon:hover {
  color: #ffffff;
  background-color: #0078d4;
}"
`;

exports[`Input should be support error 1`] = `
".FInput-root {
  font: inherit;
  width: 100%;
  border: 2px solid;
  outline: none;
  padding: 6px 12px;
  transition: background-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,border-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms;
  border-color: #dedede;
  border-radius: 2px;
}
.FInput-root:hover {
  border-color: #b5b5b5;
}
.FInput-root:active, .FInput-root:focus {
  border-color: #0078d4;
}
.FInput-root:disabled {
  color: #8c8c8c;
  cursor: not-allowed;
  pointer-events: none;
  background-color: #e9e9e9;
}
.FInput-error {
  border-color: #f44336;
}
.FInput-error:hover {
  border-color: #f44336;
}
.FInput-error:active, .FInput-error:focus {
  border-color: #f44336;
}
.FInput-wrapper {
  font: inherit;
  display: inline-block;
  position: relative;
  min-width: 200px;
}
.FInput-clearedIcon {
  top: 0;
  right: 0;
  cursor: pointer;
  height: 100%;
  display: inline-flex;
  position: absolute;
  font-size: 12px;
  transition: background-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms,border-color 250ms cubic-bezier(0.76, 0.24, 0.23, 0.94) 0ms;
  align-items: center;
  justify-content: center;
}
.FInput-clearedIcon:hover {
  color: #ffffff;
  background-color: #0078d4;
}"
`;
